File: src/client/lib/outbox.ts
// Durable outbox for unacked client ops (MG4.3).
//
// Every op the shell authors is appended here BEFORE it is sent on the wire and
// is removed only when the server acks its `opId`. This makes the unacked tail
// survive a reload/reconnect: on (re)connect the engine replays every entry in
// `localSeq` order so an op authored offline (or in flight when the socket
// dropped) is re-sent rather than lost. Re-sending the same `opId` is safe — the
// server dedups by it and re-acks the original seq (idempotent).
//
// C11: the eventual durable backing is storage-core wasm sqlite (OPFS);
// IndexedDB is the interim impl, swap behind OutboxStore.

import type { RenderScenePatch } from "../../shared/renderPatch";

/** `(clientId, localSeq)` idempotency key — mirrors the server `OpId`. */
export type OpId = {
  clientId: string;
  localSeq: number;
};

/**
 * One outbox row: an opId-stamped envelope around a whole `RenderScenePatch`.
 * Field-for-field the `ops` envelope the evolved WS protocol carries, so an
 * entry can be re-sent verbatim with no re-encoding.
 */
export type OutboxEntry = {
  opId: OpId;
  baseRevision: number;
  ts: string;
  patch: RenderScenePatch;
};

/**
 * Durable append-only log of unacked ops, keyed by `opId`.
 *
 * `append` persists before send; `remove` drops acked ids; `all` returns the
 * replay set in `localSeq` order. Implementations must keep `localSeq`
 * monotonic per `clientId` and persist that counter alongside the rows so it
 * never repeats across reloads.
 */
export interface OutboxStore {
  /** Persist an entry (call before sending it on the wire). */
  append(entry: OutboxEntry): Promise<void>;
  /** All unacked entries, ascending by `localSeq` (replay order). */
  all(): Promise<OutboxEntry[]>;
  /** Drop the entries whose `opId` is in `opIds` (on ack/rejected). */
  remove(opIds: OpId[]): Promise<void>;
  /** Drop everything (e.g. a hard reset). */
  clear(): Promise<void>;
  /** Next monotonic `localSeq` for this client; advances and persists. */
  nextLocalSeq(): Promise<number>;
}

/** True when two opIds are the same logical op. */
export function opIdEquals(a: OpId, b: OpId): boolean {
  return a.clientId === b.clientId && a.localSeq === b.localSeq;
}

/** Stable string key for an opId (Set/Map membership). */
export function opIdKey(opId: OpId): string {
  return `${opId.clientId}:${opId.localSeq}`;
}

// ---------------------------------------------------------------------------
// In-memory impl — used by tests and as a no-persistence fallback.
// ---------------------------------------------------------------------------

/**
 * Non-durable `OutboxStore` backed by a plain array. Entries are lost on reload;
 * tests simulate a "reconnect" by reusing the SAME instance (the durable case),
 * which is what the IndexedDB impl guarantees across a real reload.
 */
export class InMemoryOutboxStore implements OutboxStore {
  private entries: OutboxEntry[] = [];
  private seq = 0;

  async append(entry: OutboxEntry): Promise<void> {
    this.entries.push(entry);
  }

  async all(): Promise<OutboxEntry[]> {
    return [...this.entries].sort((a, b) => a.opId.localSeq - b.opId.localSeq);
  }

  async remove(opIds: OpId[]): Promise<void> {
    if (opIds.length === 0) return;
    const drop = new Set(opIds.map(opIdKey));
    this.entries = this.entries.filter((e) => !drop.has(opIdKey(e.opId)));
  }

  async clear(): Promise<void> {
    this.entries = [];
  }

  async nextLocalSeq(): Promise<number> {
    this.seq += 1;
    return this.seq;
  }
}

// ---------------------------------------------------------------------------
// IndexedDB impl — the interim durable backing (C11).
// ---------------------------------------------------------------------------

const DB_NAME = "shape-ai-outbox";
const STORE = "ops";
const META = "meta";

/**
 * Durable `OutboxStore` on IndexedDB. One object store keyed by `opIdKey`, plus
 * a `meta` store holding the per-client `localSeq` high-water mark so the
 * counter survives reloads. The `clientId` namespaces this client's rows so two
 * tabs sharing the DB never collide on localSeq.
 */
export class IndexedDbOutboxStore implements OutboxStore {
  private readonly clientId: string;
  private dbPromise: Promise<IDBDatabase> | null = null;

  constructor(clientId: string) {
    this.clientId = clientId;
  }

  private db(): Promise<IDBDatabase> {
    if (this.dbPromise) return this.dbPromise;
    this.dbPromise = new Promise((resolve, reject) => {
      const req = indexedDB.open(DB_NAME, 1);
      req.onupgradeneeded = () => {
        const db = req.result;
        if (!db.objectStoreNames.contains(STORE)) db.createObjectStore(STORE);
        if (!db.objectStoreNames.contains(META)) db.createObjectStore(META);
      };
      req.onsuccess = () => resolve(req.result);
      req.onerror = () => reject(req.error);
    });
    return this.dbPromise;
  }

  private async tx<T>(
    stores: string[],
    mode: IDBTransactionMode,
    run: (tx: IDBTransaction) => Promise<T> | T
  ): Promise<T> {
    const db = await this.db();
    return new Promise<T>((resolve, reject) => {
      const tx = db.transaction(stores, mode);
      let result: T;
      Promise.resolve(run(tx)).then(
        (r) => {
          result = r;
        },
        (e) => reject(e)
      );
      tx.oncomplete = () => resolve(result);
      tx.onerror = () => reject(tx.error);
      tx.onabort = () => reject(tx.error);
    });
  }

  async append(entry: OutboxEntry): Promise<void> {
    await this.tx([STORE], "readwrite", (tx) => {
      tx.objectStore(STORE).put(entry, opIdKey(entry.opId));
    });
  }

  async all(): Promise<OutboxEntry[]> {
    const rows = await this.tx<OutboxEntry[]>([STORE], "readonly", (tx) => {
      return new Promise<OutboxEntry[]>((resolve, reject) => {
        const req = tx.objectStore(STORE).getAll();
        req.onsuccess = () => resolve(req.result as OutboxEntry[]);
        req.onerror = () => reject(req.error);
      });
    });
    return rows
      .filter((r) => r.opId.clientId === this.clientId)
      .sort((a, b) => a.opId.localSeq - b.opId.localSeq);
  }

  async remove(opIds: OpId[]): Promise<void> {
    if (opIds.length === 0) return;
    await this.tx([STORE], "readwrite", (tx) => {
      const store = tx.objectStore(STORE);
      for (const id of opIds) store.delete(opIdKey(id));
    });
  }

  async clear(): Promise<void> {
    await this.tx([STORE], "readwrite", (tx) => {
      tx.objectStore(STORE).clear();
    });
  }

  async nextLocalSeq(): Promise<number> {
    const key = `localSeq:${this.clientId}`;
    return this.tx<number>([META], "readwrite", (tx) => {
      const store = tx.objectStore(META);
      return new Promise<number>((resolve, reject) => {
        const get = store.get(key);
        get.onsuccess = () => {
          const next = ((get.result as number | undefined) ?? 0) + 1;
          const put = store.put(next, key);
          put.onsuccess = () => resolve(next);
          put.onerror = () => reject(put.error);
        };
        get.onerror = () => reject(get.error);
      });
    });
  }
}

File: crates/client-runtime/src/outbox.rs
//! Durable outbox for unacked client ops.
//!
//! Every authored op is appended here BEFORE the wire send and removed only on
//! ack, so the unacked tail survives a reload/reconnect: on (re)connect the
//! engine replays every entry in `local_seq` order. Re-sending the same `op_id`
//! is safe — the server dedups by it and re-acks idempotently.
//!
//! Persistence sits behind the [`OutboxStore`] port (the web shell keeps an
//! IndexedDB impl; this crate keeps only the in-memory one). The port is
//! synchronous: append/remove/replay/seq are pure decisions, the backend's async
//! IO is the shell's concern.

pub use shape_scene_core::wire::OpId;
use shape_scene_core::wire::WireOp;

/// One outbox row: the `op_id`-stamped [`WireOp`] envelope the WS protocol
/// carries, so an entry re-sends verbatim with no re-encoding.
pub type OutboxEntry = WireOp;

/// A persistence-backend failure surfaced by an [`OutboxStore`]. The in-memory
/// impl never fails; a durable backend maps its errors onto this.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct OutboxError(pub String);

impl core::fmt::Display for OutboxError {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        write!(f, "outbox store error: {}", self.0)
    }
}

/// Stable string key for an `op_id` (set/map membership): `clientId:localSeq`.
pub fn op_id_key(op_id: &OpId) -> String {
    format!("{}:{}", op_id.client_id, op_id.local_seq)
}

/// Durable append-only log of unacked ops, keyed by `op_id`. Implementations
/// must keep `local_seq` monotonic per `client_id` and persist that counter so
/// it never repeats across reloads.
pub trait OutboxStore {
    /// Persist an entry (call before sending it on the wire).
    fn append(&mut self, entry: OutboxEntry) -> Result<(), OutboxError>;
    /// All unacked entries, ascending by `local_seq` (replay order).
    fn all(&self) -> Result<Vec<OutboxEntry>, OutboxError>;
    /// Drop the entries whose `op_id` is in `op_ids` (on ack/rejected).
    fn remove(&mut self, op_ids: &[OpId]) -> Result<(), OutboxError>;
    /// Drop everything (e.g. a hard reset).
    fn clear(&mut self) -> Result<(), OutboxError>;
    /// Next monotonic `local_seq` for this client; advances and persists.
    fn next_local_seq(&mut self) -> Result<i64, OutboxError>;
    /// Replace the contents with durable rows read back on a fresh-session
    /// reconnect, so the engine can replay them. A durable backend persists its
    /// own `local_seq` high-water, so the default only clears + re-appends; the
    /// in-memory store overrides to also reset the counter.
    fn reseed(&mut self, entries: Vec<OutboxEntry>) -> Result<(), OutboxError> {
        self.clear()?;
        for entry in entries {
            self.append(entry)?;
        }
        Ok(())
    }
}

/// Non-durable [`OutboxStore`] backed by a `Vec`. Entries are lost on reload;
/// tests simulate a "reconnect" by reusing the SAME instance (the durable case).
#[derive(Default)]
pub struct InMemoryOutboxStore {
    entries: Vec<OutboxEntry>,
    seq: i64,
}

impl InMemoryOutboxStore {
    pub fn new() -> Self {
        Self::default()
    }
}

impl OutboxStore for InMemoryOutboxStore {
    fn append(&mut self, entry: OutboxEntry) -> Result<(), OutboxError> {
        self.entries.push(entry);
        Ok(())
    }

    fn all(&self) -> Result<Vec<OutboxEntry>, OutboxError> {
        let mut out = self.entries.clone();
        out.sort_by_key(|e| e.op_id.local_seq);
        Ok(out)
    }

    fn remove(&mut self, op_ids: &[OpId]) -> Result<(), OutboxError> {
        if op_ids.is_empty() {
            return Ok(());
        }
        let drop: std::collections::HashSet<String> = op_ids.iter().map(op_id_key).collect();
        self.entries.retain(|e| !drop.contains(&op_id_key(&e.op_id)));
        Ok(())
    }

    fn clear(&mut self) -> Result<(), OutboxError> {
        self.entries.clear();
        Ok(())
    }

    fn next_local_seq(&mut self) -> Result<i64, OutboxError> {
        self.seq += 1;
        Ok(self.seq)
    }

    fn reseed(&mut self, entries: Vec<OutboxEntry>) -> Result<(), OutboxError> {
        self.seq = entries.iter().map(|e| e.op_id.local_seq).max().unwrap_or(0);
        self.entries = entries;
        Ok(())
    }
}
